samples/apps/basic/custom_endpoints/registry.h

```diff
@@ -0,0 +1,599 @@
+      extensions.emplace_back(
+        std::make_shared<ccf::js::extensions::ConsensusExtension>(this));
+      // add ccf.host.*
+      extensions.emplace_back(
+        std::make_shared<ccf::js::extensions::HostExtension>(
+          context.get_subsystem<ccf::AbstractHostProcesses>().get()));
+      // add ccf.historical.*
+      extensions.emplace_back(
+        std::make_shared<ccf::js::extensions::HistoricalExtension>(
+          &context.get_historical_state()));
+
+      interpreter_cache->set_interpreter_factory(
+        [extensions](ccf::js::TxAccess access) {
+          auto interpreter = std::make_shared<ccf::js::core::Context>(access);
+
+          for (auto extension : extensions)
+          {
+            interpreter->add_extension(extension);
+          }
+
+          return interpreter;
+        });
+
+      auto put_custom_endpoints = [this](ccf::endpoints::EndpointContext& ctx) {
+        const auto& caller_identity =
+          ctx.template get_caller<ccf::UserCOSESign1AuthnIdentity>();
+
+        // Authorization Check
+        nlohmann::json user_data = nullptr;
+        auto result =
+          get_user_data_v1(ctx.tx, caller_identity.user_id, user_data);
+        if (result == ccf::ApiResult::InternalError)
+        {
+          ctx.rpc_ctx->set_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            fmt::format(
+              "Failed to get user data for user {}: {}",
+              caller_identity.user_id,
+              ccf::api_result_to_str(result)));
+          return;
+        }
+        const auto is_admin_it = user_data.find("isAdmin");
+
+        // Not every user gets to define custom endpoints, only users with
+        // isAdmin
+        if (
+          !user_data.is_object() || is_admin_it == user_data.end() ||
+          !is_admin_it.value().get<bool>())
+        {
+          ctx.rpc_ctx->set_error(
+            HTTP_STATUS_FORBIDDEN,
+            ccf::errors::AuthorizationFailed,
+            "Only admins may access this endpoint.");
+          return;
+        }
+        // End of Authorization Check
+
+        const auto j = nlohmann::json::parse(
+          caller_identity.content.begin(), caller_identity.content.end());
+        const auto wrapper = j.get<ccf::js::BundleWrapper>();
+
+        auto endpoints = ctx.tx.template rw<ccf::endpoints::EndpointsMap>(
+          "public:custom_endpoints.metadata");
+        // Similar to set_js_app
+        for (const auto& [url, methods] : wrapper.bundle.metadata.endpoints)
+        {
+          for (const auto& [method, metadata] : methods)
+          {
+            std::string method_upper = method;
+            nonstd::to_upper(method_upper);
+            const auto key = ccf::endpoints::EndpointKey{url, method_upper};
+            endpoints->put(key, metadata);
+          }
+        }
+
+        auto modules =
+          ctx.tx.template rw<ccf::Modules>("public:custom_endpoints.modules");
```

Comment:
Can we take the table names (the string `"public:custom_endpoints.modules"`) as a constructor arg, store as a const member? Think that gives us very clear customisability already, for minimal additional code complexity.

Reply:
Yes, although I was thinking of doing that for a namespace instead of individual tables. I think the table naming remains an internal detail of the registry, maybe.